### PR TITLE
MOSIP-41175

### PR DESCRIPTION
**Pull request by kaledOu**
## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests covering client matching (exact, partial, phonetics), name/address normalization, and phonetics utilities; includes edge cases and mock isolation.

* **Chores**
  * Added build-time test tooling, coverage reporting and static-analysis (Sonar) integration, plus test-scoped Mockito dependency to improve CI quality metrics.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: demosdk/pom.xml [hunk 1]
```diff
@@ -193,6 +201,12 @@
 			<artifactId>junit</artifactId>
 			<scope>test</scope>
 		</dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>5.20.0</version>
```
Comment (JanardhanBS-SyncByte) [inline demosdk/pom.xml]: declare version on top in properties as <mockito.core.version>5.20.0</mockito.core.version>

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Test
+    public void testDoExactMatch_sameStrings_returns100() {
+        int v = client.doExactMatch("John Doe", "John Doe", new HashMap<>());
+        assertEquals("Exact match should return 100", Client_V_1_0.EXACT_MATCH_VALUE, v);
+    }
+
+    @Test
+    public void testDoExactMatch_differentOrder_returns100() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        int v = client.doExactMatch("John Doe", "Doe John", new HashMap<>());
+        assertEquals("Exact match should be order-insensitive and return 100", Client_V_1_0.EXACT_MATCH_VALUE, v);
+    }
+
+    @Test
+    public void testDoExactMatch_differentSize_returns0() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        int v = client.doExactMatch("John Doe", "John", new HashMap<>());
+        assertEquals("Different sizes should not be exact match", 0, v);
+    }
+
+    @Test
+    public void testDoPartialMatch_oneMatch_returns50() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        int v = client.doPartialMatch("John", "John Doe", new HashMap<>());
+        // matchedList.size() = 1, originalEntityInfoList.size() = 2 => 1*100/(2+0) = 50
+        assertEquals("Partial match should compute proportional value", 50, v);
+    }
+
+    @Test
+    public void testDoPartialMatch_singleCharPrefix_matchingStartsWith() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        // ref has single char token 'J' that should match entity word starting with 'J'
+        int v = client.doPartialMatch("J", "John", new HashMap<>());
+        // matchedList will be empty but unmatchedList size 1, originalEntityInfoList size 1 => 0*100/(1+0)=0
+        assertEquals(0, v);
+    }
+
+    @Test
+    public void testDoPartialMatch_singleCharPrefix_foundMatch_returns50() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        int v = client.doPartialMatch("J Doe", "John Doe", new HashMap<>());
+        // matchedList.size() = 1 (Doe), originalEntityInfoList.size() = 2 => 1*100/(2+0) = 50
+        assertEquals(50, v);
+    }
+
+    @Test
+    public void testDoExactMatch_sameSize_but_notAllMatch_returns0() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        int v = client.doExactMatch("A B", "A C", new HashMap<>());
+        assertEquals(0, v);
+    }
+
+    @Test
+    public void testDoExactMatch_withNulls_returns100() {
+        // passing empty strings instead of null to avoid NPE from production code
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java:83]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        int v = client.doExactMatch("", "", new HashMap<>());
+        assertEquals(Client_V_1_0.EXACT_MATCH_VALUE, v);
+    }
+
+    @Test
+    public void testDoPhoneticsMatch_callsTextMatcherUtil() throws EncoderException {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        mockedStatic.when(() -> TextMatcherUtil.phoneticsMatch("abc", "abc", "en")).thenReturn(100);
+        int val = client.doPhoneticsMatch("abc", "abc", "en", new HashMap<>());
+        assertEquals(100, val);
+    }
+
+    @Test
+    public void testDoPhoneticsMatch_handlesEncoderException() throws Exception {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/ClientV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            java.lang.reflect.Field f = Normalizer_V_1_0.class.getDeclaredField("environment");
+            f.setAccessible(true);
+            f.set(normalizer, env);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void testRemoveTitlesAndNormalizeName_basic() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+        // no normalization patterns configured
+        when(env.getProperty(anyString())).thenReturn(null);
+
+        String out = normalizer.normalizeName("Mr John Doe", "en", titles);
+        assertEquals("John Doe", out);
+    }
+
+    @Test
+    public void testNormalizeAddress_withPatternReplacement() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+import org.junit.Before;
+import org.junit.Test;
+import org.springframework.core.env.Environment;
+
+import static org.mockito.Mockito.*;
+
+public class NormalizerV1UnitTest {
+
+    private Normalizer_V_1_0 normalizer;
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java:21]: NormalizerV10
PascalCase

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+        String out = normalizer.normalizeAddress("123 Main St", "en");
+        // digits replaced by # => "# Main St"
+        assertTrue(out.startsWith("#"));
+    }
+
+
+
+    @Test
+    public void testGetBasicNormalisers_multiplePatterns() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+            Map<Pattern, String> map = (Map<Pattern, String>) m.invoke(normalizer, "ida.demo.name.normalization.regex.en[%s]");
+            // expect both patterns present
+            assertEquals(2, map.size());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void testGetBasicNormalisers_address_emptyReplacement_and_normalizeAddress() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+            // digits removed -> should contain 'Main'
+            assertTrue(out.contains("Main") || out.contains("main"));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void testNormalizeName_with_common_and_any_patterns() {
+        // configure patterns across name and common to ensure normalizeWithCommonAttributes merges them
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java:109]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+        when(env.getProperty("ida.norm.sep", "=")).thenReturn("=");
+
+        Map<String, List<String>> titles = new HashMap<>();
+        String out = normalizer.normalizeName("john doe", "en", titles);
+        // both replacements applied
+        assertTrue(out.contains("J") || out.contains("D"));
+    }
+
+    @Test
+    public void testNormalizeName_removeAllCases_variousCasingsAndPunctuation() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+        // create a name with multiple occurrences, different casings and punctuation
+        String input = "Mr. JOHN Dr mr DR. Mr";
+        String out = normalizer.normalizeName(input, "en", titles);
+        // all title tokens should be removed, leaving only JOHN
+        assertTrue(out.trim().equalsIgnoreCase("john") || out.trim().contains("john"));
+    }
+
+    @Test
+    public void testNormalizeName_removeAllCases_repeatedOccurrences() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+
+        String input = "Hon Hon. HON honHon Hon.";
+        String out = normalizer.normalizeName(input, "en", titles);
+        // after removing Hon variants, result should be 'honHon' normalized to keep unmatched sequence
+        String normalized = out.trim().toLowerCase();
+        assertTrue("Should not contain 'hon' as separate token", !normalized.matches(".*\\bhon\\b.*"));
+    }
+
+    @Test
+    public void testNormalize_handlesZeroLengthRegexMatch_breaksLoop() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+        when(env.getProperty("ida.norm.sep", "=")).thenReturn("=");
+
+        Map<String, List<String>> titles = new HashMap<>();
+        String out = normalizer.normalizeName("abc", "en", titles);
+        // zero-length match case: implementation breaks and should not loop infinitely; output should be trimmed string
+        assertTrue(out != null && out.length() >= 0);
+    }
+
+    @Test
+    public void testRemoveAllCases_coversOriginalLowerUpperIndexes() {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/impl/spec_1_0/NormalizerV1UnitTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/utils/TextMatcherUtilTest.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+import static org.junit.Assert.*;
+
+import org.apache.commons.codec.EncoderException;
+import org.junit.Test;
+
+public class TextMatcherUtilTest {
+
+    @Test
+    public void testPhoneticsMatch_identicalStrings_returnsPositiveMultipleOf20() throws EncoderException {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/utils/TextMatcherUtilTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/utils/TextMatcherUtilTest.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+        Integer v = TextMatcherUtil.phoneticsMatch("John", "John", "english");
+        assertNotNull(v);
+        assertTrue(v > 0);
+        assertEquals(0, v.intValue() % 20);
+        assertTrue(v >= 20 && v <= 100);
+    }
+
+    @Test
+    public void testPhoneticsMatch_differentStrings_returnsValueInExpectedRange() throws EncoderException {
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/utils/TextMatcherUtilTest.java]: java camel case

File: demosdk/src/test/java/io/mosip/demosdk/client/utils/TextMatcherUtilTest.java [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+        Integer v2 = TextMatcherUtil.phoneticsMatch("John", "Doe", "english");
+        assertNotNull(v1);
+        assertNotNull(v2);
+        assertEquals(0, v1.intValue() % 20);
+        assertEquals(0, v2.intValue() % 20);
+        assertTrue(v1 >= 20 && v1 <= 100);
+        assertTrue(v2 >= 20 && v2 <= 100);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
```
Comment (JanardhanBS-SyncByte) [inline demosdk/src/test/java/io/mosip/demosdk/client/utils/TextMatcherUtilTest.java:31]: java camel case

**Review by JanardhanBS-SyncByte — CHANGES_REQUESTED** [review]

**Review by JanardhanBS-SyncByte — APPROVED** [review]

**Review by Prafulrakhade — APPROVED** [review]
